### ISSUE TITLE
Release a beta of the Firefox WebExtensions port

### ISSUE THREAD

**Issue by Bisaloo** [Synzvato/decentraleyes#173]
If I got it right, the webextension port is ready or close to be since you released a chromium version of decentraleyes.

I think you should release a beta before the official release of Firefox 57. HTTPS-Everywhere just did this and the firefox and chromium API are apparently different enough to cause some issues.

I believe many followers of this github repo would be happy to help you test it.

**Comment by zymase** [comment]
No hurry. WebExtension is by itself a pain in the neck, moreover it won't be mandatory before Firefox 57, next November. Hence, November 2017 is the only milestone. otherwise, while we're at it, why not ask developers to update their add-ons to Nightly builds? Well, some users do, some users even score badly an ad-on because it's not alpha-compatible! Hey: carpe diem! one thing at a time, please, mercy!

**Comment by heubergen** [comment]
I think FF57 Beta should be targed as goal to port this extensions to WE and this would be the 2017-09-26.

**Comment by sabret00the** [comment]
56 would be better as the target (basically what's on Nightly now) so that there's enough time for testing and catching any bugs.

**Comment by welwood08** [comment]
If the configuration of existing legacy users is to be kept and imported to any webext version, the only way to do that is with a hybrid version that embeds a transitional webext to copy existing settings into webext-accessible storage. This must be done before FF57 (because the hybrid won't be able to load at all in FF57).

**Comment by Bisaloo** [comment]
I think some of the reasoning here are flawed. The FF and Chromium WebExtensions implementation are still pretty different. It is very likely that there will be bugs to fix.

*What we don't know* is how much time it is going to take to fix them. So the safest bet is to release the beta as soon as possible. The sooner bugs are reported, the longer @Synzvato (or anyone willing to help) will have to fix them.

**Comment by zymase** [comment]
> [...]we may possibly end up with no working decentraleyes when FF 57 comes out and it would be a shame.
That would be no working decentraleyes on Firefox. There are alternatives to Firefox. I have in mind Waterfox if it manages to carry on legacy add-ons after Firefox 56.

Decentraleyes is a major add-on, it is for me together with a few others, not for cosmetics but for privacy and security. Should any of those add-ons not be available (sometimes because not technically feasible) that I'll move to a Firefox fork. And I won't be the only one.

**Comment by mrdokenny** [comment]
https://addons.mozilla.org/en-US/firefox/addon/local-cdn-webextension/

I tested this and in my limited testing, it has some problems of its own, but someone else might have better luck.

**Comment by heubergen** [comment]
@mrdokenny  Thanks for the effort but if you just repacked the chrome sources I can tell you that your not the first one that tried that :) If there more I think @Synzvato would be more than happy to merge it.
The Extensions seems not to work at all (https://decentraleyes.org/test/) so I think you can just delete it again :)

**Comment by mrdokenny** [comment]
@heubergen I think you are under the impression that this is my extension, but it's not. It's something I found on the mega list of WebExtension alternatives.

https://www.reddit.com/r/firefox/comments/6i1fu2/webext_equivalents_to_legacy_addons/

**Comment by heubergen** [comment]
Oh sorry, my bad.

**Comment by Synzvato** [comment]
@mrdokenny @heubergen Thank you for looking into this fork, but it is indeed broken. It has since been removed from the list of ```WebExtensions``` equivalents, and is no longer publicly listed on ```AMO```.

I'm actively working on porting the add-on over to Firefox 57, but doing so is more complicated than I had hoped. I'm aiming for a ```stable``` release on September 1<sup>st</sup>. If said release deadline cannot be met, Mozilla will no longer longer be marking this extension as featured on ```AMO```. It all depends on the possibilities.

The first ```alpha``` (or ```beta```) will be made available as soon as there a working prototype. I'm happy to see that there's a lot of interest in this release, and I'll be sure to keep you all posted. Any suggestions or ideas are welcome! Please know that the development of the ```WebExtensions``` port has top priority.

**Comment by Synzvato** [comment]
The September 1<sup>st</sup> deadline is now less than one month away, but things are looking up. I'm convinced a ```WebExtensions``` port is feasible, and expect to be able to release the beta within the next two weeks.

**Comment by sabret00the** [comment]
Thank you for all of your hard work and effort @Synzvato. I would recommend that you keep an eye on https://bugzilla.mozilla.org/show_bug.cgi?id=1347507 [as at least one of the module owners of Mozilla has said they're going to be aggressively ripping out XUL compatible components as of yesterday](https://www.reddit.com/r/firefox/comments/6qdm1q/nightly_57_is_coming_nightly_57_is_coming/dkxb7yc/).

**Comment by ronjouch** [comment]
Today August 11 is the day for users running Nightly: [PSA: Legacy extensions disabled by default on Nightly](https://mail.mozilla.org/pipermail/dev-addons/2017-August/003059.html).

**Comment by gitarra** [comment]
@ronjouch This addon was already broken since version bump to 57.

**Comment by ronjouch** [comment]
No it wasn't, I was using Decentraleyes until today. (And they're still usable if you flip `extensions.legacy.enabled` in `about:config`). I was just mentioning the mailing-list announcement to share the news of the moving train 🙂.

**Comment by darnir** [comment]
@gitarra It indeed was still working. In fact, I don't have today's nightly and Decentraleyes is working for me.

Since this is the very last legacy addon I have, it would be really nice if a useable beta was already out for us to convert.

**Comment by Synzvato** [comment]
Thanks, @ronjouch, @gitarra, and @darnir. Does the legacy extension pass the [testing utility](https://decentraleyes.org/test) checks? Expect a Firefox 56 and 57 compatible ```WebExtensions``` beta within the next few days.

**Comment by ronjouch** [comment]
@Synzvato

> *"Does the legacy extension pass the [testing utility](https://decentraleyes.org/test) checks?"*

Nope. With the legacy extension installed (manually enabled through `extensions.legacy.enabled`) and a restart, the test util says:

```
Your browser was able to connect to a blacklisted CDN.
Decentraleyes is not working as intended.
```

> *"Expect a Firefox 56 and 57 compatible WebExtensions beta within the next few days."*

Awesome! Thanks for your work!

**Comment by darnir** [comment]
For me it does pass the utility check.
I have Decentraleyes along with uMatrix running.

**Comment by sabret00the** [comment]
It doesn't pass for me on three mobile devices and a two different laptop profiles running 57.

**Comment by Synzvato** [comment]
@darnir @sabret00the It shouldn't be passing any tests within Firefox 57, as a couple of essential legacy APIs have been stripped from its codebase. So, this is where things end for the legacy add-on.

**Comment by Synzvato** [comment]
### Before Proceeding

I have both good and bad news with regards to the new ```WebExtensions``` beta. The good news is that ```v2.0.0-beta1``` is fully finished, and that the corresponding source code has been pushed into the ```master``` branch of this repository. The legacy extension's code can be found inside of the ```legacy``` branch.

The bad news is that due to a bug in Firefox 56 and Firefox 57, the extension only works properly when it's unpacked. I have reported these [findings](https://bugzilla.mozilla.org/show_bug.cgi?id=1256122#c103), and this resulted in the discovery of bug [1390346](https://bugzilla.mozilla.org/show_bug.cgi?id=1390346). It's being actively worked on, and I'm convinced the issue will be fixed before Firefox 56 turns stable.

Since an official, signed, beta release would surely confuse people into thinking the extension rewrite is fundamentally broken, I am unable to release a packaged version of the first beta at this point.

---

So, with all that out of the way, let's dive into how you can start using the beta today.

### Get the Code

- Download the code inside of the master branch to your computer.
- Remove the ```.git``` folder from the resulting collection of files and folders.
- Add the rest to a zip archive, and change its file extension to ```.xpi```.

### Install the Extension

_Important: The guides below propose disabling signature verification. Only do
so if you accept the risks involved with trusting unsigned add-ons._

#### For Desktop Users

- Make sure that you are using a Firefox version that allows you to disable
signature verification and is equal to or higher than 56 (e.g. Nightly).
- Navigate to ```about:config```, change ```xpinstall.signatures.required``` to
```false```, and then set ```extensions.alwaysUnpack``` to ```true```.
- Open the new ```.xpi``` archive with Firefox, and follow the instructions.

_Note: Do not forget to revert any preference changes, once you switch over
to official builds of the extension, as these are temporary workarounds._

#### For Mobile Users

- Make sure that you are using a Firefox version that allows you to disable
signature verification and is equal to or higher than 56 (e.g. Nightly).
- Visit ```about:config```, and then change ```xpinstall.signatures.required```
to ```false```. That's all, because Fennec is not affected by the bug.
- Read these [instructions](https://developer.mozilla.org/en-US/Add-ons/WebExtensions/Developing_WebExtensions_for_Firefox_for_Android#Install_and_run_your_extension_in_Firefox_for_Android) to learn more about installing custom extensions
on mobile devices. Pick the approach you feel most comfortable with.

_Note: Do not forget to revert any preference changes, once you switch over
to official builds of the extension, as these are temporary workarounds._

---

Special thanks go to Mozilla's Rob Wu, for his valuable support during the porting procedure. Also, many thanks to everyone close to me, who supported me on a personal level.

I would also like to thank all contributors and users that keep this project alive, and am looking forward to hearing your thoughts on the new codebase. Please don't hesitate to leave feedback!

**Comment by EchoDev** [comment]
Thanks for all the work @Synzvato 

I wonder if the packed XPI version would work on stable (FF55). Does FF55 have all the API's needed for the webextension version to work?

Last time I checked you could easily sign extensions on AMO if you handle the distribution on your own. I might give it a try if FF55 is able to use the webextension version.

**Comment by ronjouch** [comment]
w00t 👍. I confirm 2.0.0-beta1 installs fine on Nightly 57.0a1 (2017-08-16) with the above procedure and prefs, and the test page confirms "Decentraleyes is fully operational".

**Comment by t-kitsune** [comment]
Thank you so much for this!

**Comment by Radagast** [comment]
I haven't had a chance to install the webextension version yet, however, on Nightly, the legacy version still reports being "Fully Operational" with `extensions.legacy.enabled`set to true. 

I'll try and get to the webext later today.

**Comment by Radagast** [comment]
Update: Webextension seems to be working fine.

Quick question, are there any rule updates for uMatrix with this release?

Thank you for this :)

**Comment by Synzvato** [comment]
> Unless I did something wrong it doesn't seem to work FF55. The extension does load, but it fails the test on the Decentraleyes homepage. (tested packed and unpacked)

@EchoDev You're very welcome, and thanks for sharing your personal findings! Firefox 55 does indeed lack some cutting edge features that this ```WebExtensions``` port needs to properly inject resources.

> I confirm 2.0.0-beta1 installs fine on Nightly 57.0a1 (2017-08-16) with the above procedure and prefs, and the test page confirms "Decentraleyes is fully operational".

@ronjouch Great to hear that it's completely working on your side, and thanks for sharing!

> Thank you so much for this!

@t-kitsune My pleasure, thanks for trying out the beta!

> I haven't had a chance to install the webextension version yet, however, on Nightly, the legacy version still reports being "Fully Operational" with ```extensions.legacy.enabled``` set to true.

@Radagast That's interesting. Perhaps some breaking changes only affect specific platforms.

> Webextension seems to be working fine [...], are there any rule updates for uMatrix [...]?

@Radagast Good to hear. Decentraleyes serves as a last layer of defense for requests that would otherwise be sent off to a Content Delivery Network. So, when using this extension alongside uMatrix (or any other kind of content blocker), just stick with the policies you feel most comfortable with.

Since the FAQ-entry on this is a tad outdated, confusing, and only explains how to loosen your current blocking policies (to be able to configure blocking rules through Decentraleyes), I will be sure to rewrite it before this port turns stable. In fact, I'll be revising a lot of the existing documentation.

>  [...] Thank you for this :)

@Radagast You're welcome, many thanks for the feedback!

**Comment by Bisaloo** [comment]
I could install it without any issues on nightly following your instructions. It works fine according to the testing utility.

However, it looks like DNS requests for the CDN (`ajax.googleapis.com` in this case) are still going through. I can't remember if this was already the case before and if this is known/intended behaviour.

**Comment by Synzvato** [comment]
@Bisaloo Even if this were the case, these requests would not end up at the CDN in question, but at the DNS provider you already entrust with lookup of the actual domain you're visiting.

Please see related issue #98 for more details, and feel free to weigh in on the discussion.

**Comment by Bisaloo** [comment]
@Synzvato, after looking at it more, I think it's more than DNS requests going through. I sent you an email. Let me know if you can reproduce.

**Comment by Synzvato** [comment]
> [...] after looking at it more, I think it's more than DNS requests going through.

@Bisaloo Please create a new issue if you think your web browser is sending out more than just DNS requests, so we can continue this particular conversation there.

**Comment by sabret00the** [comment]
The Firefox bug that blocked the release of this on the developement channel on AMO landed and so as from the next Nightly release, the packed version will work. 

See the Bugzilla bug for more details: https://bugzilla.mozilla.org/show_bug.cgi?id=1390346

**Comment by Synzvato** [comment]
> [...] I sent you an email. Let me know if you can reproduce.

@Bisaloo I have created a [new issue](https://github.com/Synzvato/decentraleyes/issues/188) that's dedicated to your findings. It contains detailed information on the bug, and explains the recently implemented [solution](https://github.com/Synzvato/decentraleyes/commit/7d21a1ed924517c3041925da2a82d90a078983fc). Thanks again for sharing your concerns.

This issue affects any early adopters of the beta who have network prediction set to enabled, and are not using content blockers (e.g. uBlock Origin) that automatically disable this feature by default.

**Comment by tukoz** [comment]
@Synzvato thank you for the continous efforts you're putting it; I bet you deserve the « stakanovitch » badge amongst FF add-ons I follw at least ;)

Question: On Nightly 57.0a1 i686 after I followed the [instructions above](#issuecomment-322821857), FF throws me a « Add-on corrupted; can't install ». Some one please help me find what I'm doing wrong? Have downloaded the .zip archive from the project Master and changed its extension to `.xpi`. Thank you for a possible fix to this.

**Comment by pwd-github** [comment]
So the [Mozilla bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1390346) has now been marked fixed meaning that no unpacking needs to be done. So you shouldn't have to worry any more @tukoz as I'm sure that when @Synzvato gets time, he'll upload the WebExtension to AMO and it should work out of the box.

**Comment by Synzvato** [comment]
> [...] thank you for the continous efforts you're putting it; I bet you deserve the « stakanovitch » badge amongst FF add-ons I follw at least ;)

@tukoz Thanks, hehe. I'm very happy to hear you appreciate the work that's going into this!

> Some one please help [...] Have downloaded the .zip archive from the project Master and changed its extension to ```.xpi```. Thank you for a possible fix to this.

When you download the code as a ```.zip``` archive, the contents reside inside in a subfolder. Make sure that the actual project files are in the root folder of your extension package. I hope this helps!

> So the [...] bug has now been marked fixed [...]. I'm sure that when [the author] gets time, he'll upload the WebExtension to AMO and it should work out of the box.

@pwd-github Thanks for keeping everyone updated! Although the fix has indeed landed, it's currently set to be included in Firefox 57. We're currently [attempting](https://bugzilla.mozilla.org/show_bug.cgi?id=1390346#c40) to get it uplifted to Firefox 56.

I'm actively looking into this, and a packaged version will appear as soon as absolutely possible.

**Comment by pwd-github** [comment]
@Synzvato I noticed that you're not using a development channel on AMO for Decentraleyes. Any chance you can upload the WebExtension to the development channel so that us Nightly users can use it while you await the uplift to 56?

**Comment by Synzvato** [comment]
> I noticed that you're not using a development channel on AMO [...]. Any chance you can upload the WebExtension to the development channel [...] while you await the uplift to 56?

@pwd-github Once I can confirm that the Nightly update that fixes the issue has been distributed, I'll be uploading ```beta-1``` to AMO. The beta tag will cause the channel to be created automatically.

**Comment by pwd-github** [comment]
Thank you @Synzvato, I and I expect everyone else appreciates all of your efforts.

**Comment by Synzvato** [comment]
Decentraleyes ```v2.0.0beta1``` is [now available](https://addons.mozilla.org/firefox/addon/decentraleyes/versions/beta?page=1#version-2.0.0beta1) inside of the AMO development channel. Alternatively, you can download the package from the corresponding [release page](https://github.com/Synzvato/decentraleyes/releases/tag/v2.0.0beta1) in this repository.

**Comment by pwd-github** [comment]
Well done and thank you @synzvato.

**Comment by hobarrera** [comment]
Odd, it claims to be incompatible with FF 55, but I've a few other WebExtension plugins running on FF 55. Is this an oversight, or does it actually depend on newer FF features?

**Comment by Synzvato** [comment]
> Well done and thank you [...].

@pwd-github Thank you, and I'm glad you like the outcome!

> Odd, it claims to be incompatible with FF 55, but I've a few other WebExtension plugins running on FF 55. Is this an oversight, or does it actually depend on newer FF features?

@hobarrera Although it might look as if Decentraleyes is unnecessarily late to the party, the new ```WebExtension``` does indeed depend on features that are not present in Firefox 55. Luckily, Mozilla has been kind enough [to uplift](https://bugzilla.mozilla.org/show_bug.cgi?id=1390346#c44) some required fixes to Firefox 56. They will likely land in ```56.0b7```.

It's definitely not an oversight, and ```v2.0``` will land as early as absolutely possible!

**Comment by hobarrera** [comment]
Ah, thanks for the clarification! I was assuming it might be this.

**Comment by savyajha** [comment]
@Synzvato Since FF 56.0b7 has landed, do you think it's okay to allow the webext to install on FF 56? I tried installing the packaged version from github, and I get the message that it's not supported on 56. :(

**Comment by Synzvato** [comment]
@savyajha Good question! I would recommend waiting just a tad longer. Decentraleyes ```v2.0.0beta2``` will be targeting Firefox 56 _(and higher)_, and should be ready sometime next week.

**Comment by davidhedlund** [comment]
@Synzvato Is Decentraleyes a hybrid WebExtension now?

**Comment by Synzvato** [comment]
@davidhedlund Yes, ```v1.3.10``` uses an embedded WebExtension for migrating preferences.

**Comment by davidhedlund** [comment]
@Synzvato Thank you.

An add-on is a hybrid WE if it has a "hasEmbeddedWebExtension" key available in the install.rdf

I didn't know that was the standard procedure when I asked, sorry.

**Comment by Synzvato** [comment]
Decentraleyes ```v2.0.0beta2``` is [now available](https://addons.mozilla.org/firefox/addon/decentraleyes/versions/beta?page=1#version-2.0.0beta2) inside of the AMO development channel. Alternatively, you can download the package from the corresponding [release page](https://github.com/Synzvato/decentraleyes/releases/tag/v2.0.0beta2) inside of this repository.

_Note: This beta release is compatible with Firefox_ ```56.0b7``` _(and higher)._

**Comment by andschwa** [comment]
The pre-release version 2.0.0beta3 works with Firefox Beta (57.0b3); however, for some reason AMO still says:

> This add-on is not compatible with your version of Firefox.

I am speaking specifically about version 2.0.0beta3 under the "Development Channel." But downloading it and installing manually worked anyway.

**Comment by majewsky** [comment]
@Synzvato Do I understand correctly, that 2.0.0 will become stable when Firefox 56 is out in the wild, and users on Stable won't have to do anything else to prepare for 57 (other than auto-updating to Decentraleyes 2.0.0)?

**Comment by Synzvato** [comment]
> The pre-release version 2.0.0beta3 works with Firefox Beta (57.0b3); however, for some reason AMO still says: This add-on is not compatible [...] on my macOS machine.

@andschwa Thanks for reporting this! I have double-checked the compatibility metadata, but couldn't find any issues there. It's explicitly marked as being compatible with "All Platforms". I'll be sure to ask other macOS users to have a go at reproducing the problem. I'll post additional details if anything is found.

> Thank you for this wonderful extension!

@andschwa You're welcome, and I'm glad to hear you like it!

> Do I understand correctly, that 2.0.0 will become stable when Firefox 56 is out in the wild, and users on Stable won't have to do anything else to prepare for 57 [...]?

@majewsky That's fully correct! As soon as enough people use Firefox 56, the ```WebExtension``` will be released, and existing installations can be (auto-)updated. Make sure you're using the latest version of the ```legacy``` extension before upgrading to ```v2.0.0```  in case you want your settings to be migrated over.

**Comment by savyajha** [comment]
@andschwa Looks like you might have `privacy.resistFingerprinting` set to true on your Mac.

**Comment by andschwa** [comment]
@savyajha oh, hey! Look at that. I guess it's working as intended 😄

**Comment by OdinGitDat** [comment]
> That's fully correct! As soon as most people use Firefox 56, the WebExtension will be released.

56 has been rolled out for a few days now.

**Comment by Synzvato** [comment]
> 56 has been rolled out for a few days now.

Yes, I do know, but thanks for making sure! Since ```v2.0.0``` is incompatible with Firefox 55, I am waiting a couple of extra days, to keep the majority of (potential) users from running into installation issues.

**Comment by Bisaloo** [comment]
I have no idea how https://addons.mozilla.org/ works. Out of curiosity, what will happen to Firefox ESR users?

**Comment by evaryont** [comment]
They will download the latest version that is marked compatible, like any other Firefox. If they don't update to the next version of ESR, which will include WebExtension support, they will most likely be stuck at version is compatible. (Individual developers may or may not continue development of the two different types of extensions, but that is hard so I don't expect so.)

**Comment by Bisaloo** [comment]
> They will download the latest version that is marked compatible, like any other Firefox

In that case, what's the harm in uploading decentraleyes `v2.0.0` now as long as it is marked incompatible with FF < 56?

I feel like I'm missing something.

**Comment by Synzvato** [comment]
> I don't think there is any plan to continue development of the not-WebExtension version? So I think the latest compatible version is the 1.x series, and not 2.0.

@evaryont There is a plan to provide the ```legacy``` extension with security updates, libraries, and other small improvements. However, both extensions will remain completely separated. So yes, ```1.x.x```.

> In that case, what's the harm in uploading decentraleyes v2.0.0 now as long as it is marked incompatible with FF < 56? [...] I feel like I'm missing something.

@Bisaloo The problem is, that Firefox 55 or ESR users would not directly see the older version. They would be greeted with a multitude of warnings, as well as a grayed out download button.

![image](https://user-images.githubusercontent.com/2366020/31317051-33f6b160-ac10-11e7-8d9c-ae978601045f.png)
![image](https://user-images.githubusercontent.com/2366020/31317053-39cd105c-ac10-11e7-819e-16866af598ec.png)

**Comment by Synzvato** [comment]
The stable release of Decentraleyes ```v2.0.0``` (for Firefox) is [now available](https://addons.mozilla.org/en-US/firefox/addon/decentraleyes/) on AMO. Alternatively, you can get the extension from the corresponding [release page](https://github.com/Synzvato/decentraleyes/releases/tag/v2.0.0) inside of this repository. Thanks for waiting!

**Comment by collinbarrett** [comment]
awesome, thanks. it does say it is not compatible for me with Firefox Developer Edition on Windows. but, that could be tied to the Dev. Edition being a version or two ahead?

![image](https://user-images.githubusercontent.com/6483057/31516892-aa170d02-af5f-11e7-8fa4-cdbfefd783d3.png)

**Comment by welwood08** [comment]
Excellent news!

@collinbarrett Dev Edition on Windows is fine, I have seen that happen when using certain privacy preferences (such as `privacy.resistFingerprinting` which disguises your browser version to a fixed value).

**Comment by collinbarrett** [comment]
@welwood08 Ah. Thanks. That's exactly what it is. I am using https://github.com/pyllyukko/user.js/
